okapi-core/src/main/java/org/folio/okapi/util/ModuleVersionFilter.java

```diff
@@ -0,0 +1,71 @@
+package org.folio.okapi.util;
+
+import io.vertx.core.MultiMap;
+import io.vertx.core.json.DecodeException;
+import java.util.Optional;
+import org.folio.okapi.common.ModuleId;
+
+public class ModuleVersionFilter {
+  Boolean npmSnapshot;
+  Boolean preRelease;
+
+  public ModuleVersionFilter() {
+    this(null, null);
+  }
+
+  ModuleVersionFilter(Boolean preRelease, Boolean npmSnapshot) {
+    this.preRelease = preRelease;
+    this.npmSnapshot = npmSnapshot;
+  }
+
+  /**
+   * Construct from HTTP query parameters "npmSnapshot", "preRelease".
+   * @param params HTTP parameters
+   */
+  public ModuleVersionFilter(MultiMap params) {
+    Optional<Boolean> o = getParamVersionFilter(params, "npmSnapshot");
+    npmSnapshot = o.isEmpty() ? null : o.get();
+
+    o = getParamVersionFilter(params, "preRelease");
+    preRelease = o.isEmpty() ? null : o.get();
+  }
+
+  /**
+   * Check if module ID matches preRelease/npmSnapshot spec.
+   * @param idThis module ID
+   * @return true if module should be included; false it should be filtered away.
+   */
+  public boolean matchesModule(ModuleId idThis) {
+    if (idThis.hasPreRelease()) {
+      return preRelease == null || preRelease;
+    } else if (idThis.hasNpmSnapshot()) {
+      return npmSnapshot == null || npmSnapshot;
+    } else {
+      return ((npmSnapshot == null || !npmSnapshot)
+          && (preRelease == null || !preRelease));
+    }
+  }
+
+  /**
+   * Lookup boolean preRelease/npmSnapshot parameter in HTTP request.
+   * @param params HTTP server request parameters
+   * @param name name of query parameter
+   * @return returns false for "false", true for "only", and
+   *     Optional.empty() for "true" and null/undefined
+   * @throws DecodeException on invalid value
+   */
+  static Optional<Boolean> getParamVersionFilter(MultiMap params, String name) {
```

Comment:
Can you also replace `Optional<Boolean>` by `Boolean` here?

Reply:
ok. I'll try.. Just note that SonarQube warned when I returned null earlier..

Reply:
I tried it again #1184. So I'll keep the  optional and use orElse.

Reply:
SonarQube's warning is a false positive and should be suppressed. The method is used in this class only and the value is intended to be one of three states, one of them is null. There is no danger that someone forgets that the value might be null.

Using `Optional<Boolean> v` with `v.get()` may throw NPE, using `Boolean v` and accessing `v` may also throw NPE. This is exactly the same danger.

If you consider SonarQube's warning about the method to be valid then this assignment to a Boolean
`npmSnapshot = getParamVersionFilter(params, "npmSnapshot").orElse(null)`
is equally dangerous.